governance/program/src/state/proposal.rs

```diff
@@ -15,51 +25,273 @@ pub struct Proposal {
+            return Err(GovernanceError::InvalidStateCannotAddSignatory.into());
+        }
+
+        Ok(())
+    }
+    /// Checks if Signatory can be removed from the Proposal in the given state
+    pub fn assert_can_remove_signatory(&self) -> Result<(), ProgramError> {
+        if !(self.state == ProposalState::Draft || self.state == ProposalState::SigningOff) {
+            return Err(GovernanceError::InvalidStateCannotRemoveSignatory.into());
+        }
+
+        Ok(())
+    }
+
+    /// Checks if Proposal can be singed off
+    pub fn assert_can_sign_off(&self) -> Result<(), ProgramError> {
+        if self.state != ProposalState::Draft && self.state != ProposalState::SigningOff {
+            return Err(GovernanceError::InvalidStateCannotSignOff.into());
+        }
+
+        if self.signatories_count == 0 {
+            return Err(GovernanceError::ProposalHasNoSignatories.into());
+        }
+
+        Ok(())
+    }
+}
+
+/// Deserializes Proposal account and checks owner program
+pub fn deserialize_proposal_raw(proposal_info: &AccountInfo) -> Result<Proposal, ProgramError> {
+    deserialize_account::<Proposal>(proposal_info, &id())
+}
+
+/// Returns Proposal PDA seeds
+pub fn get_proposal_address_seeds<'a>(
+    governance: &'a Pubkey,
+    governing_token_mint: &'a Pubkey,
+    name: &'a str,
+) -> [&'a [u8]; 4] {
+    [
+        PROGRAM_AUTHORITY_SEED,
+        governance.as_ref(),
+        governing_token_mint.as_ref(),
+        &name.as_bytes(),
+    ]
+}
+
+/// Returns Proposal PDA address
+pub fn get_proposal_address<'a>(
+    governance: &'a Pubkey,
+    governing_token_mint: &'a Pubkey,
+    name: &'a str,
+) -> Pubkey {
+    Pubkey::find_program_address(
+        &get_proposal_address_seeds(governance, governing_token_mint, name),
+        &id(),
+    )
+    .0
+}
+
+#[cfg(test)]
+mod test {
+
+    use {super::*, proptest::prelude::*};
+
+    fn create_test_proposal() -> Proposal {
+        Proposal {
+            account_type: GovernanceAccountType::TokenOwnerRecord,
+            governance: Pubkey::new_unique(),
+            governing_token_mint: Pubkey::new_unique(),
+            state: ProposalState::Draft,
+            token_owner_record: Pubkey::new_unique(),
+            signatories_count: 10,
+            signatories_signed_off_count: 5,
+            description_link: "This is my description".to_string(),
+            name: "This is my name".to_string(),
+            draft_at: 10,
+            signing_off_at: Some(10),
+            voting_at: Some(10),
+            voting_completed_at: Some(10),
+            executing_at: Some(10),
+            closed_at: Some(10),
+            number_of_executed_instructions: 10,
+            number_of_instructions: 10,
+        }
+    }
+
+    #[test]
+    fn test_max_size() {
+        let proposal = create_test_proposal();
+        let size = proposal.try_to_vec().unwrap().len();
+
+        assert_eq!(proposal.get_max_size(), Some(size));
```

Comment:
No, it doesn't limit name and description to ~20 characters, because it's a `max_size` of an instance of the account and not `max_size` for the account type.

```
impl AccountMaxSize for Proposal {
    fn get_max_size(&self) -> Option<usize> {
        Some(self.name.len() + self.description_link.len() + 163)
    }
}

```
 It uses dynamic size for name and description and extra space for the fields which can grow. The test  guards the optional (resizable) fields are counted for. For example the Options in the test are set to Some to give max size.

Reply:
My apologies, you're right.  To be honest, I was confused about the `AccountMaxSize` trait.  Makes sense then!